designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,440 @@
+
+#. Support one dependency manifest (ex: package.json) entire app or one per each Lambda function.
+
+#. Support out-of-source builds: ie. source code of Lambda function is outside the directory containing SAM template.
+
+#. Integrate build action with ``sam local/package/deploy`` commands so the Lambda functions will be automatically
+   built as part of the command without explicitly running the build command.
+
+
+Out-of-Scope
+------------
+#. Supports adding data files ie. files that are not referenced by the package manager (ex: images, css etc)
+#. Support to exclude certain files from the built artifact (ex: using .gitignore or using regex)
+#. If the app contains a ``buildspec.yaml``, automatically run it using CodeBuild Local.
+#. Watch for file changes and build automatically (ex: ``sam build --watch``)
+#. Support other build systems by default Webpack, Yarn or Gradle.
+#. Support in AWS CLI, ``aws cloudformation``, suite of commands
+#. Support for fine-grained hooks (ex: hooks that run pre-build, post-build, etc)
+
+
+User Experience Walkthrough
+---------------------------
+Let's assume customers has the following SAM template:
+
+    **NOTE**: Currently we advice customers to set *CodeUri* to a folder containing built artifacts that can be readily
+    packaged by ``sam package`` command. But to use with *build* command, customers need to set *CodeUri* to the folder
+    containing source code and not built artifacts.
+
+.. code-block:: yaml
+
+    MyFunction1:
+        Type: AWS::Lambda::Function
+        Properties:
+            ...
+            Code: ./source-code1
+            ...
+
+    MyFunction2:
+        Type: AWS::Serverless::Function
+        Properties:
+            ...
+            Code: ./source-code2
+            ...
+
+
+To build, package and deploy this app, customers would do the following:
+
+**1. Build:** Run the following command to build all functions in the template and output a SAM template that can be run through
+the package command:
+
+.. code-block:: bash
+
+    # Build the code and write artifacts to ./build folder
+    $ sam build -t template.yaml -b ./build -o built-template.yaml
+
+
+Output of the *sam build* command is a SAM template where CodeUri is pointing to the built artifacts. Note the values of
+Code properties in following output:
+
+.. code-block:: bash
+
+    $ cat built-template.yaml
+    MyFunction1:
+        Type: AWS::Lambda::Function
+        Properties:
+            ...
+            Code: ./build/MyFunction1
+            ...
+
+    MyFunction2:
+        Type: AWS::Serverless::Function
+        Properties:
+            ...
+            CodeUri: ./build/MyFunction2
+            ...
+
+**2. Package and Deploy:** Package the built artifacts by running the *package* command on the template output
+by *build* command
+
+.. code-block:: bash
+
+    # Package the code
+    $ sam package --template-file built-template.yaml --s3-bucket mybucket --output-template-file packaged-template.yaml
```

Comment:
Similar to previous comment on UX, we could share the vision as to how defaults would be used here:

``sam package --s3-bucket mybucket``. ``--output-template-file`` and ``template-file`` should have these as defaults

Reply:
I don't want to change `sam package` functionality now. We should do that eventually when we migrate package completely into SAM CLI